### PR TITLE
Add tests to restart the docker and verify volume refcounts are restored

### PR DESCRIPTION
**Pull request by govint**
Newly added tests are passing except for the failing scenario of #1499. Most likely #1499 is happening in each of the newly added tests  that are restarting containers post docker restart.

Test result is shown below and the failure for which #1499 is created. This branch can be used to repro #1499 as needed.

Test result:
2017/06/28 09:25:50 Removing container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:25:51 Creating volume [restart_test_volume_602275] on VM [10.161.250.171]
2017/06/28 09:25:54 Adding access for sharedVmfs-1 to vmgroup _DEFAULT  with max-size 1gb, total-size 100gb
2017/06/28 09:25:55 Restart tests setup complete
2017/06/28 09:25:55 START: RestartTestData.TestDuplicateVolumeName
2017/06/28 09:25:55 Attaching volume [restart_test_volume_602275] on VM[10.161.250.171]
2017/06/28 09:25:57 Attaching volume [restart_test_volume_602275@sharedVmfs-1] on VM[10.161.250.171]
2017/06/28 09:26:03 Restarting docker ....
2017/06/28 09:26:37 Attaching volume [restart_test_volume_602275@sharedVmfs-0] on VM[10.161.250.171]
2017/06/28 09:26:45 Removing container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:26:46 Removing container [restart_test_container_901653] on VM [10.161.250.171]
2017/06/28 09:26:46 Removing container [restart_test_container_578687] on VM [10.161.250.171]
2017/06/28 09:26:47 Attaching volume [restart_test_volume_602275] on VM [10.161.248.164]
2017/06/28 09:27:19 Destroying volume [restart_test_volume_602275@sharedVmfs-1]
2017/06/28 09:27:20 END: RestartTestData.TestDuplicateVolumeName
2017/06/28 09:27:20 START: RestartTestData.TestLongVolumeName
2017/06/28 09:27:20 Attaching volume [restart_test_volume_602275] on VM[10.161.250.171]
2017/06/28 09:27:21 Restarting docker ....
2017/06/28 09:27:37 Attaching volume [restart_test_volume_602275] on VM[10.161.250.171]
2017/06/28 09:27:45 Attaching volume [restart_test_volume_602275@sharedVmfs-0] on VM[10.161.250.171]
2017/06/28 09:27:45 Removing container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:27:45 Removing container [restart_test_container_901653] on VM [10.161.250.171]
2017/06/28 09:27:46 Removing container [restart_test_container_578687] on VM [10.161.250.171]
2017/06/28 09:27:47 Attaching volume [restart_test_volume_602275] on VM [10.161.248.164]
2017/06/28 09:27:49 END: RestartTestData.TestLongVolumeName
2017/06/28 09:27:49 START: RestartTestData.TestPluginKill
2017/06/28 09:27:49 Attaching volume [restart_test_volume_602275] on VM[10.161.250.171]
2017/06/28 09:27:50 Confirming attached status for volume [restart_test_volume_602275]
2017/06/28 09:27:53 Killing vDVS plugin on VM [10.161.250.171]
2017/06/28 09:27:54 Sleep for 2 seconds
2017/06/28 09:27:56 Stopping container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:28:07 Starting container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:28:08 Stopping container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:28:19 Confirming detached status for volume [restart_test_volume_602275]
2017/06/28 09:28:19 Sleep for 2 seconds
2017/06/28 09:28:23 Removing container [restart_test_container_499262] on VM [10.161.250.171]
2017/06/28 09:28:23 END: RestartTestData.TestPluginKill
2017/06/28 09:28:23 START: RestartTestData.TestRecoverMountsAfterRestart
2017/06/28 09:28:23 Attaching volume [restart_test_volume_602275] on VM[10.161.250.171]
2017/06/28 09:28:24 Attaching volume [restart_test_volume_602275] on VM[10.161.250.171]
2017/06/28 09:28:25 Restarting docker ....
2017/06/28 09:28:41 Sleep for 60 seconds

----------------------------------------------------------------------
FAIL: restart_test.go:190: RestartTestData.TestRecoverMountsAfterRestart

restart_test.go:210:
    c.Assert(s.config.DockerHostNames[0], Equals, status["attached_vm"], Commentf("Volume is not attached to VM %s, attached to %s",
        s.config.DockerHostNames[0], status["attached_vm"]))
... obtained string = "worker1-VM1.0"
... expected string = "<no"
... Volume is not attached to VM worker1-VM1.0, attached to <no

2017/06/28 09:29:42 START: RestartTestData.TestVolumeDetached
2017/06/28 09:29:42 Attaching volume [restart_test_volume_602275] on VM [10.161.250.171]

### CODE REVIEW

**Comment by govint** [comment]
I'd suggest skipping the failing test and once #1499 is fixed, reenable it.

**Comment by govint** [comment]
Have skipped the one test that's addressed by #1499. The remaining tests can be reviewed.

**Comment by pshahzeb** [comment]
@govint 
The description of your PR says TestRecoverMountsAfterRestart fails for which we have #1499.
However that test seems to pass on CI. 
But it failed at a different test case :-) .

Could you please confirm if it is a test issue or a bug ? 
From the test steps in TestDuplicateVolumeName, it looks like there is a bug.


```
2017/06/28 20:31:02 Destroying volume [restart_test_volume_666978@vsanDatastore]

----------------------------------------------------------------------
FAIL: restart_test.go:291: RestartTestData.TestDuplicateVolumeName

restart_test.go:337:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc420196e80), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: unable to remove volume: remove restart_test_volume_666978@vsanDatastore: VolumeDriver.Remove: Remove failure - volume is still mounted.  volume=restart_test_volume_666978@vsanDatastore, refcount=1

2017/06/28 20:31:03 START: RestartTestData.TestLongVolumeName
```

File: tests/e2e/restart_test.go [hunk 1]
```diff
@@ -21,39 +21,59 @@ package e2e
...
 	"github.com/vmware/docker-volume-vsphere/tests/utils/misc"
+	adminconst "github.com/vmware/docker-volume-vsphere/tests/constants/admincli"
+	adminutil "github.com/vmware/docker-volume-vsphere/tests/utils/admincli"
 )
 
 type RestartTestData struct {
 	config        *inputparams.TestConfig
 	volumeName    string
-	containerName string
+	containerName []string
```
Comment (pshahzeb) [inline tests/e2e/restart_test.go]: nit: might wanna call it containerNameList or containerList

File: tests/e2e/restart_test.go [hunk 2]
```diff
@@ -21,39 +21,59 @@ package e2e
...
 
 	s.volumeName = inputparams.GetUniqueVolumeName("restart_test")
-	s.containerName = inputparams.GetUniqueContainerName("restart_test")
+	s.containerName = []string{inputparams.GetUniqueContainerName("restart_test"),
+				inputparams.GetUniqueContainerName("restart_test"),
+				inputparams.GetUniqueContainerName("restart_test")}
 
 	// ensure there are no remanants from an earlier run
-	out, err := dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerName)
+	out, err := dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerName[0])
```
Comment (pshahzeb) [inline tests/e2e/restart_test.go]: why just ensure the first container name? looks like we have 3.

File: tests/e2e/restart_test.go [hunk 3]
```diff
@@ -21,39 +21,59 @@ package e2e
...
 
 	// Create a volume
 	out, err = dockercli.CreateVolume(s.config.DockerHosts[1], s.volumeName)
 	c.Assert(err, IsNil, Commentf(out))
+
+	// Add access for the second datastore to the VM's vmgroup
+	out, err = adminutil.AddDatastoreToVmgroup(s.config.EsxHost, adminconst.DefaultVMgroup, s.config.Datastores[1])
```
Comment (pshahzeb) [inline tests/e2e/restart_test.go]: Here we don't know config has two datastores.

input param valids a test config if it has atleast one datastore, not two. Quoting the check below
```
	if config.DockerHostNames[0] == "" || config.DockerHostNames[1] == "" ||
		len(config.Datastores) <= 1 {
		return nil
	}
```

**Comment by pshahzeb** [comment]
Oh I missed the 
```
c.Skip("Skipping verification of attached status after docker restart")
```

File: tests/e2e/restart_test.go [hunk 4]
```diff
@@ -137,23 +157,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+
+	// 1. Run container with restart-always flag
+	out, err := dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerName[0])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 2. Restart docker
+	out, err = dockercli.RestartDocker(s.config.DockerHosts[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.SleepForSec(60)
```
Comment (pshahzeb) [inline tests/e2e/restart_test.go]: What happens if we don't sleep and immediately fire another attach here?

**Comment by govint** [comment]
@pshahzeb, yes there is a bug in how volumes are identified during refcount init.

Its like this, suppose we have two volumes of the same name (thats a real good test for duplicate names - 👍 ) on different datastores.

On startup after a restart,
1. Docker restarts the containers, but the mounts only mark state as dirty and don't increment the respective counters.
1. Refcnt init runs and gets the containers with mounted volumes and handles each one.
1. We call GetVolumeInfo() which returns the full name of the volume and refcnt module increments the counters.
1. The bug here is in GetVolumeInfo() we look at the refmap and compare the volume names alone. In GetNameFromRefmap(). Since we have two volumes with the same name on different datastores we can (in one scenario) end up double counting a single volume. 

The fix is to first get the full name of the volume and in GetNameFromRefmap() compare the full names and not the volume name alone.

From the logs in CI,
1. Two containers are restarted by docker and they mount two volumes of the same name on different datastores.
2017-06-28 20:29:49.626432838 +0000 UTC [INFO] Mounting volume name="restart_test_volume_666978" 
2017-06-28 20:29:49.891919721 +0000 UTC [INFO] Mounting volume name="restart_test_volume_666978@vsanDatastore" 

2. Refcount init has run and it has double counted one of the volumes, the result is the double counted volume remains stuck with the VM (not detached because of extra refcount) and the other volume is unmounted even though the volume is used by a container.

2017-06-28 20:30:08.881354165 +0000 UTC [INFO] Getting volume data from unix:///var/run/docker.sock
2017-06-28 20:30:08.892820898 +0000 UTC [INFO] Found 2 running or paused containers
2017-06-28 20:30:08.89565598 +0000 UTC [INFO] Initiating recovery unmount. dev="/dev/disk/by-path/pci-0000:0b:00.0-scsi-0:0:0:0" name="restart_test_volume_666978@datastore1" refcnt=0 mounted=true 
2017-06-28 20:30:09.827914074 +0000 UTC [INFO] Discovered 2 volumes in use.

_This volume is double counted_
2017-06-28 20:30:09.828002239 +0000 UTC [INFO] Volume name=restart_test_volume_666978@vsanDatastore count=2 mounted=true device='/dev/disk/by-path/pci-0000:0b:00.0-scsi-0:0:1:0'

_This volume is not counted at all and hence unmounted above._
2017-06-28 20:30:09.828014114 +0000 UTC [INFO] Volume name=restart_test_volume_666978@datastore1 count=0 mounted=true

File: tests/e2e/restart_test.go (hunk 4 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Without the sleep (this may need documentation or a rethink on refcount init), the remove container calls translate to unmount requests from docker to the plugin. The plugin is not ready yet so we mark dirty on the refcount and the unmount returns immediately.

The refcounter in parallel may or may not have found the running containers and can keep a reference count on the volume causing it to remain attached to the VM. Which is why the earlier CI failed (without the sleep).

At least unmount should fail the call to docker if the refcounter is still not initialized perhaps.

File: tests/e2e/restart_test.go (hunk 1 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Done.

File: tests/e2e/restart_test.go (hunk 2 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Fixed.

File: tests/e2e/restart_test.go (hunk 3 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Fixed.

**Comment by govint** [comment]
Killing the docker daemon stops the containers and when docker restarts it restarts the containers. The plugin is also restarted as a result of docker restarting. Which should result in the same scenario as if the docker is restarted.

**Comment by pshahzeb** [comment]
> Killing the docker daemon stops the containers and when docker restarts it restarts the containers. 
Correct.
However, killing docker kills containers and the last volume isn't actually unmounted after all containers are killed (reason being we missed unmounts)
In restarting, docker first stops all the containers.. we get respective unmounts and we cleanly unmount volume when the last container is stop.

This makes a difference when to the state after the plugin comes up. This essentially hits a different set of code paths.

**Comment by govint** [comment]
Then I'll add a kill docker test after this set is merged.

File: tests/e2e/restart_test.go [hunk 5]
```diff
@@ -137,23 +159,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+
+	// 3. Verify the volume is attached to the VM
+	status, err := dockercli.GetVolumeStatus(s.config.DockerHosts[1], s.volumeName)
+	c.Assert(err, IsNil, Commentf("Failed to fetch status for volume %s", s.volumeName))
+	c.Assert(s.config.DockerHostNames[0], Equals, status["attached_vm"], Commentf("Volume is not attached to VM %s, attached to %s",
```
Comment (pshahzeb) [inline tests/e2e/restart_test.go]: This assert compares the attached_vm status to s.config.DockerHostNames[0].
But lines 197 and 201 attach the container from s.config.DockerHosts[1]. These are different hosts.
How is this assert going to pass?  Am I missing anything here?

Comment (govint) [inline tests/e2e/restart_test.go]: Why not? Docker volume status is the same from any node. That's a test in itself.

Comment (govint) [inline tests/e2e/restart_test.go]: Good catch, checked the CI and this test didn't run and why???

2017/06/29 09:40:01 END: RestartTestData.TestPluginKill
2017/06/29 09:40:01 START: RestartTestData.TestRecoverMountsAfterRestart
2017/06/29 09:40:01 START: RestartTestData.TestVolumeDetached
2017/06/29 09:40:01 Attaching volume [restart_test_volume_1037905] on VM [192.168.31.85]
2017/06/29 09:40:03 Confirming attached status for volume [restart_test_volume_1037905]

Comment (pshahzeb) [inline tests/e2e/restart_test.go]: because of this line 195? 
```
	c.Skip("Skipping verification of attached status after docker restart")
```

Comment (govint) [inline tests/e2e/restart_test.go]: ha, my turn now

**Comment by govint** [comment]
Make of plugin failed

mkdir -p /tmp/plugin/rootfs
docker export  tempContainer  | tar -x -C /tmp/plugin/rootfs
sh: 0: getcwd() failed: No such file or directory
cp config.json docker-volume-vsphere /tmp/plugin
cp: cannot stat 'config.json': No such file or directory
cp: cannot stat 'docker-volume-vsphere': No such file or directory
make: *** [plugin] Error 1
Makefile:95: recipe for target 'plugin' failed
scp: /tmp/docker-volume-vsphere/build: No such file or directory
sh: 0: Can't open /tmp/docker-volume-vsphere/build/plugin_sanity_test.sh
bash: line 0: cd: /tmp/docker-volume-vsphere/plugin: No such file or directory
make: *** No rule to make target 'push'.  Stop.

File: tests/e2e/restart_test.go (hunk 5 above)
Comment (pshahzeb) [inline tests/e2e/restart_test.go]: This skip still exists in latest change on this PR. As per #1499, it's a corner case. Should this skip be removed so that this test runs ?

**Comment by pshahzeb** [comment]
testcase TestRecoverMountsAfterRestart is still skipped. Should we allow it?

File: tests/utils/admincli/vmgroupmgmt.go [hunk 6]
```diff
@@ -195,3 +195,10 @@ func GetDBmode(esxIP string) string {
 		return ""
 	}
 }
+
+// RemvoeDatastoreAccessFromVmgroup - Remove access for a datastore from a vmgroup
+func RemoveDatastoreFromVmgroup(ip, vmgroup, dsName string) (string, error) {
+	log.Printf("Removing access for %s to vmgroup %s", dsName, vmgroup)
```
Comment (shaominchen) [inline tests/utils/admincli/vmgroupmgmt.go]: Nit: to => from

File: tests/e2e/restart_test.go [hunk 7]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 2. Restart docker
+	out, err = dockercli.RestartDocker(s.config.DockerHosts[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.SleepForSec(60)
+
+	// 3. Verify the volume is attached to the VM
+	status, err := dockercli.GetVolumeStatus(s.config.DockerHosts[1], s.volumeName)
```
Comment (shaominchen) [inline tests/e2e/restart_test.go]: Why don't we use existing VerifyAttachedStatus util function here?

File: tests/e2e/restart_test.go (hunk 4 above)
Comment (shaominchen) [inline tests/e2e/restart_test.go]: Can we do polling in a loop to check the status? Blind sleeping is inefficient and sometimes unstable.

File: tests/e2e/restart_test.go [hunk 8]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+
+	// 2. Run second container with same volume name on the other datastore 
+	out, err = dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName + "@" + ds2, s.containerNameList[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 3. Restart docker
+	out, err = dockercli.RestartDocker(s.config.DockerHosts[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.SleepForSec(60)
```
Comment (shaominchen) [inline tests/e2e/restart_test.go]: Same as above.

File: tests/e2e/restart_test.go [hunk 9]
```diff
@@ -21,39 +21,68 @@ package e2e
...
-	if s.config == nil {
-		c.Skip("Unable to retrieve test config, skipping docker/plugin restart tests")
+	if s.config == nil || len(s.config.Datastores) < 2 {
+		log.Printf("Restart tests setup skipped - missing or incomplete config")
+		c.Skip("Restart tests setup skipped - missing or incomplete config")
 	}
 
-	s.volumeName = inputparams.GetUniqueVolumeName("restart_test")
-	s.containerName = inputparams.GetUniqueContainerName("restart_test")
+	adminutil.ConfigRemove(s.config.EsxHost)
```
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: this is redundant as per https://github.com/vmware/docker-volume-vsphere/blob/master/tests/utils/admincli/vmgroupmgmt.go#L88

File: tests/e2e/restart_test.go [hunk 10]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+// 1. Start 2 containers using same volume with restart=always
+// 2. Restart docker and wait for containers to start and reference counts to be initialized.
+// 3. Verify volume is attached to the VM
+// 4. Start 1 container using same volume
+// 5. Stop all 3 containers and confirm disk is detached
+// 6. Run container on other host to verify the volume is detached after stopping the containers
+func (s *RestartTestData) TestRecoverMountsAfterRestart(c *C) {
+	misc.LogTestStart(c.TestName())
+
+	c.Skip("Skipping verification of attached status after docker restart")
```
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: shouldn't it be removed?

File: tests/e2e/restart_test.go [hunk 11]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+
+	// Run second container 
+	out, err = dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerNameList[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 2. Restart docker
+	out, err = dockercli.RestartDocker(s.config.DockerHosts[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.SleepForSec(60)
```
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: can we improve this further by using some short of poller?

File: tests/e2e/restart_test.go [hunk 12]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+	c.Assert(err, IsNil, Commentf("Failed to fetch status for volume %s", s.volumeName))
+	c.Assert(s.config.DockerHostNames[1], Equals, status["attached_vm"], Commentf("Volume is not attached to VM %s, attached to %s",
+		s.config.DockerHostNames[1], status["attached_vm"]))
+
+	// 4. Run third container
+	out, err = dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerNameList[2])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 5. Stop the three containers
+	out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerNameList[0])
```
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: L#228 to L#234 can be optimized further by calling `dockercli.RemoveContainer` only once by appending s.containerNameList[0-2] with spaces between them.

File: tests/e2e/restart_test.go [hunk 13]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+	c.Assert(err, IsNil, Commentf(out))
+
+	out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerNameList[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerNameList[2])
+	c.Assert(err, IsNil, Commentf(out))
```
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: as per step `// 5. Stop all 3 containers and confirm disk is detached` ... `confirm disk is detached` verification step is missing.

File: tests/e2e/restart_test.go [hunk 14]
```diff
@@ -137,23 +166,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
...
+
+	// 6. Run container on other host to verify the volume is detached after stopping the containers
```
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: can you please help me to understand about step#6? .. what is the objective of the same?

File: tests/utils/inputparams/testparams.go [hunk 15]
```diff
@@ -136,7 +136,6 @@ func GetTestConfig() *TestConfig {
 	config.DockerHostNames = append(config.DockerHostNames, esx.RetrieveVMNameFromIP(config.DockerHosts[0]))
 	config.DockerHostNames = append(config.DockerHostNames, esx.RetrieveVMNameFromIP(config.DockerHosts[1]))
 	config.Datastores = esx.GetDatastoreList()
-
```
Comment (shuklanirdesh82) [inline tests/utils/inputparams/testparams.go]: I guess this file can be removed from the PR.

File: tests/utils/admincli/vmgroupmgmt.go (hunk 6 above)
Comment (govint) [inline tests/utils/admincli/vmgroupmgmt.go]: Done

File: tests/e2e/restart_test.go (hunk 7 above)
Comment (govint) [inline tests/e2e/restart_test.go]: I'd added the GetVolumeStatus and decided to use that as well here. But other functions are using the verification function and I'm reverting to do that.

File: tests/e2e/restart_test.go (hunk 9 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Fixed.

File: tests/e2e/restart_test.go (hunk 10 above)
Comment (govint) [inline tests/e2e/restart_test.go]: I can't test on my set up with this test, but I'll open it up for CI.

File: tests/e2e/restart_test.go (hunk 12 above)
Comment (govint) [inline tests/e2e/restart_test.go]: RemoveContainer is taking a string arg, can't pass a list to it.
func RemoveContainer(ip, containerName string) (string, error)

File: tests/e2e/restart_test.go (hunk 13 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Correct, fixed.

File: tests/e2e/restart_test.go (hunk 14 above)
Comment (govint) [inline tests/e2e/restart_test.go]: Being able to launch a container using the same volume on another VM confirms that the volume is certainly not in the VMX config of the original VM. VerifyDetachedStatus() is using the admin CLI and KV to figure the volume is detached. These can't be trusted. Launching the container on another host gets confirmation from ESX that the volume is detached from the first VM.

File: tests/utils/inputparams/testparams.go (hunk 15 above)
Comment (govint) [inline tests/utils/inputparams/testparams.go]: Fixed

**Comment by govint** [comment]
One test (TestDuplicateVolumeNames) is failing thats passing locally, it passed earlier in CI too. Debugging.

File: tests/e2e/restart_test.go (hunk 12 above)
Comment (shuklanirdesh82) [inline tests/e2e/restart_test.go]: sorry for the confusion it might have cause. 

What I meant instead of invoking 3 calls how about following code block?

```
continerNames := s.containerNameList[0] + " " + s.containerNameList[1] + " "  + s.containerNameList[2] 
out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], continerNames)
c.Assert(err, IsNil, Commentf(out))
```

Comment (govint) [inline tests/e2e/restart_test.go]: Ok, thats not exactly how RemoveContainer is defined but it should allow to be used this way.

**Comment by shuklanirdesh82** [comment]
CI is passed and it has 2 LGTM .. merging this PR.
